AP_HAL_Linux: support PWM input for BH hat

onceme opened the pull request:
This patch enables up to 8 PWMs RCInput for RPI1/2/3 by DMA and PCM DREQ. Current code already supported 1 PPMSum RCInput, since 8 PWMs could spare extra PPM encoder hardware, we would like to introduced it.

This patch was launched before, please refer to
- https://github.com/ArduPilot/ardupilot/issues/3509
- https://github.com/ArduPilot/ardupilot/pull/3489
- https://github.com/ArduPilot/ardupilot/issues/3414

I modified and done aging/monkey test in our own fork here: https://github.com/bithollow/bitforge/issues/61

So far FS-TH9X & FS-R9B & BH board & RPI 2/3 was tested, both 1 PPMSum with extra encoder and 8 PWMs.

The performance of RCInput_RPI is increased around 8 times, memory footprint reduced to about 8% compare to original, from 3680 KB to 272 KB.

The crash issue in previous PR was found because DMA stopped, then a nullptr was derferenced.

onceme commented:
@lucasdemarchi 

Hi Lucas,
It's been a long time, I'm picking this up again:)
Could you kindly help me review the patch?

It will be really appreciated if you can confirm the patch on any RCInput_RPI enabled board such as NAVIO, ERLEBRAIN2, DARK, URUS, PXFMINI.

Greetings,

ranqingfa commented:
Great work! qq511306337 Like your bithollow!

lucasdemarchi commented:
@onceme sorry about the delay. 

The RCInput   on RPI gave me so much headache in the past that I'm trying to avoid touching it :). What we really need is to use process separation for this part of the code so if it crashes at least we don't lose the drone.  This has nothing to do with this PR, though.

So... do you have a PPM receiver to test on your board? Does this code work on BH if you define it to use PPM?  It would be good to leave it running for some hours to make sure we don't regress.

@patrickpoirier51 can you give this a try on your RPI ?

onceme commented:
@lucasdemarchi We are not hurry, just take your time would be okay:)

> The RCInput on RPI gave me so much headache in the past that I'm trying to avoid touching it :). What we really need is to use process separation for this part of the code so if it crashes at least we don't lose the drone. This has nothing to do with this PR, though.

It happens even without my past patches right? As in this comment before: https://github.com/ArduPilot/ardupilot/pull/3489#issuecomment-170924995, after reverted my past patch, crash still happened, with lower reproducing chances.

Believe me it also gave me headache in the past 2 months, I checked the bcm2835 datasheet over and over again for each DMA/PCM debug/control register, bit by bit. However, I think their settings are correct.

The thing is, I do found two points that could cause RCInput_RPI crash. One is line 487 in current latest code, when "ad" is nullptr, the code hasn't check nullptr before dereferencing it as "ad+j"; The other one, is a magic outside RCInput_RPI, it somehow corrupted the DMA control block in memory, thus DMA transfer stopped, then "ad" becomes nullptr. Please also check this comment: https://github.com/bithollow/bitforge/issues/61#issuecomment-321716523, I summarized my findings.

> So... do you have a PPM receiver to test on your board? Does this code work on BH if you define it to use PPM? It would be good to leave it running for some hours to make sure we don't regress.

Yes, as previously mentioned in this PR, I tested it with a PPM encoder (not PPM receiver though, actually PPM encoder + PWM receiver), with PPMSum input to BH. The aging test was over night, no crash and the RCInput responded well in APMPlanner GCS next morning. For the PWM test, it is still ongoing while I'm typing, works well so far, has been 12 days.

But, still, I only confirmed it with BH, if you or other buddies have other boards use RCInput_RPI, it's better asking them to try on their own. Thanks for asking!

kobylyanskiy commented:
@onceme 

Hi, Aaron,
I've tested your patch on NAVIO and got these error messages:
![image](https://user-images.githubusercontent.com/6507706/30268923-f2922784-96ef-11e7-8e4a-7ac60de03fdd.png)
I can't even connect to the GCS. I believe that something should be fixed to get it working.

onceme commented:
@kobylyanskiy Hi, thank you for testing this patch for me:), ah, firstly, I think those print messages are harmless, dropping samples are because we don't want to stay in _timer_tick() too long, that would cause delay. The logic was from the original, just added a print log to show how many samples were dropped.

Secondly, could you let me know your code base commit id? I would like to try it tomorrow my self. RCInput should not affect GCS, for now I can't tell why you couldn't connect to GCS, are you sure the ip address is correct? Before applying this patch, were you able to connect GCS?

onceme commented:
@kobylyanskiy 

Hi, Kirill,

Today I tired ardurover, arducopter, arduplane, ardusub on master branch, based on the commit:

> commit 58e784024f0b5f79a68573dadb9b63f12bf5236c
> Date:   Tue Aug 1 13:44:32 2017 -0700

By using BH board, everything works as expected, all of them can connect to GCS (APM Planner 2).

I don't have a NAVIO, so could you help me to try the attached binary on your board? This binary was built on the same code base, but waf configured as **navio**. If you still have problem, please leave msg here. Thanks in advance.

[ardurover-navio.zip](https://github.com/ArduPilot/ardupilot/files/1294981/ardurover-navio.zip)

I can even use the NAVIO version binary on BH to connect to GCS, but since board is different, several things don't work.

The attach screenshot is BH built binaries tried on my side:

![screenshot at 2017-09-12 14 08 49](https://user-images.githubusercontent.com/5754278/30312136-c8cf2cc2-97ca-11e7-8284-71ca8aca03f4.png)

onceme commented:
@lucasdemarchi 

Hi Lucas,

I found that ardusub doesn't respond to RCInput_RPI, is this normal? Although I didn't look into the code, arducopter/plane/rover runs well, just guess it is not a problem right? Thanks!

onceme commented:
@OXINARF Thanks:) Got it! Is it reasonable to make Sub RCInput available? Probably it concerns the receiver waterproofing, the radio strange under water?

OXINARF commented:
@onceme It was purposely removed so I doubt that Sub maintainer wants it back - it adds code complexity and the benefit is probably near zero for now.

lucasdemarchi commented:
> @kobylyanskiy Hi, thank you for testing this patch for me:), ah, firstly, I think those print messages are harmless, dropping samples are because we don't want to stay in _timer_tick() too long, that would cause delay. The logic was from the original, just added a print log to show how many samples were dropped.

This print should be a debug() call that's only enabled when DEBUG is defined. See how it's done on other drivers.

I'm actually a little bit worried about this message... it looks like it loses lots of samples.

lucasdemarchi commented on libraries/AP_HAL_Linux/RCInput_RPI.cpp:
```diff
@@ -476,13 +498,23 @@ void RCInput_RPI::init()
+        init_ctrl_data();
+        init_PCM();
+        init_DMA();
+        return;
```
Can you split this commit into things that are exclusive to the PWM and things like the above that are more a safety thing?  This `if` here could be applied very easily, and possibly other parts. Thanks.

onceme replied:
I divided this commit into three, still, the PWM part changed many lines. I'm trying to make it as easy to review as it can be. If you think something still can be split, please let me know.

onceme commented:
@lucasdemarchi 

Hi, Lucas

> This print should be a debug() call that's only enabled when DEBUG is defined. See how it's done on other drivers.

I replaced printf() with debug() while DEBUG is defined in https://github.com/ArduPilot/ardupilot/pull/6755/commits/ac84a9c6582790ca6d9d030be2770fc028f7b51b, is it correct? There are various kinds of debug() in the source code, I didn't see one could work with "waf configure --board xxx --debug". If there's one, please tell me:)

> I'm actually a little bit worried about this message... it looks like it loses lots of samples.

In this new commit, I also adjusted the RCIN_RPI_MAX_COUNTER, so few sample would be dropped. In my actual test, even many samples were dropped, the performance looks fine in APM Planner. I guess the DMA DREQ rate is high enough, thus samples in the circle_buffer are always synced.

Still, if you have other RPI based boards to test, please feedback to me, thanks!

lucasdemarchi commented:
I applied the fixes on this PR. The commit adding support for PWM I will delay until once I can test it.

lucasdemarchi commented:
@onceme any news on this? This doesn't seem to be going anywhere and currently I this rcinput crashes on me after just a few minutes.

onceme commented:
@lucasdemarchi Hi, recently I didn’t follow this one. Could you give me details about your test? Like sw version and what board did you use (w/wo the patch)?

We use this version:
https://bithollow.github.io/downloads/ardupilot-prebuilt/2017/08/11/ardupilot_v3.5-dev_copter/
https://bithollow.github.io/downloads/rpi-sd-image/2016/01/09/sd.image_v1.0.0/

And BH board here: 
https://bithollow.com/portfolio/bh-pilot/

It works fine so far. I haven’t think about to isolate it into another process though.

lucasdemarchi commented:
@onceme nice to see the board :)

The problems I was referring were in a erlebrain2 without these patches. I remember seeing it on a navio, too, but I'm only with a navio2 right now (that has a dedicated microcontroller).

> We use this version:
> https://bithollow.github.io/downloads/ardupilot-prebuilt/2017/08/11/ardupilot_v3.5-dev_copter/

Do you have a git tree/branch for that? What does that version has that is not in upstream? Could you check on master branch in your board if it's working?

lucasdemarchi commented:
@onceme so I tested on a navio+ without this patch and left it running for ~15h without a problem. So I think there's some problem either on the RPI that comes with the erlebrain, the kernel or something else.

> Could you have a try on this BH-develop branch? I'm on a trip at the moment, will be back on next Thursday, I could check on latest master after that.

If it continues working on the navio I think we can merge it.

> BTW, we would like to document BH Pilot information onto ardupilot.org website, like other supported boards, do you think it's possbile?

Sure, I think this is important... I didn't even know the board had materialized. I think the first thing to do is to get your name in the README.md file as the maintainer for this board so I know who to ask for help when problems/changes arrive.

As for the documentation site, send pull requests to https://github.com/ArduPilot/ardupilot_wiki - maybe @rmackay9 can help you with that.

onceme commented:
@lucasdemarchi 

Hi Lucas,

I'm checking this PR, travis-ci build failed, I think it has something wrong, [here the log](https://travis-ci.org/ArduPilot/ardupilot/jobs/405713631) says:

> +'[' '!' -d /home/travis/opt/gcc-arm-none-eabi-4_9-2015q3 ']'
> +wget http://firmware.ardupilot.org/Tools/PX4-tools/gcc-arm-none-eabi-4_9-2015q3-20150921-linux.tar.bz2
> --2018-07-19 08:56:59--  http://firmware.ardupilot.org/Tools/PX4-tools/gcc-arm-none-eabi-4_9-2015q3-20150921-linux.tar.bz2
> Resolving firmware.ardupilot.org (firmware.ardupilot.org)... 208.68.38.241
> Connecting to firmware.ardupilot.org (firmware.ardupilot.org)|208.68.38.241|:80... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2018-07-19 08:56:59 ERROR 404: Not Found.
> The command "Tools/scripts/configure-ci.sh" failed and exited with 8 during .
> Your build has been stopped.

Would you mind take a look? Thanks!

auturgy commented:
That’s a problem we’ve caused (someone changed PX4-Tools to STM32-Tools without fixing links). Should be fixed soon.

Regards,

James 

> On 20 Jul 2018, at 9:54 am, Aaron Wang Shi <notifications@github.com> wrote:
> 
> @lucasdemarchi
> 
> Hi Lucas,
> 
> I'm checking this PR, travis-ci build failed, I think it has something wrong, here the log says:
> 
> +'[' '!' -d /home/travis/opt/gcc-arm-none-eabi-4_9-2015q3 ']'
> +wget http://firmware.ardupilot.org/Tools/PX4-tools/gcc-arm-none-eabi-4_9-2015q3-20150921-linux.tar.bz2
> --2018-07-19 08:56:59-- http://firmware.ardupilot.org/Tools/PX4-tools/gcc-arm-none-eabi-4_9-2015q3-20150921-linux.tar.bz2
> Resolving firmware.ardupilot.org (firmware.ardupilot.org)... 208.68.38.241
> Connecting to firmware.ardupilot.org (firmware.ardupilot.org)|208.68.38.241|:80... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2018-07-19 08:56:59 ERROR 404: Not Found.
> The command "Tools/scripts/configure-ci.sh" failed and exited with 8 during .
> Your build has been stopped.
> 
> Would you mind take a look? Thanks!
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

onceme commented:
@lucasdemarchi 
Sorry for the late reply.
> so I tested on a navio+ without this patch and left it running for ~15h without a problem. So I think there's some problem either on the RPI that comes with the erlebrain, the kernel or something else.

It might be, but I don't know. So now you have a navio+?

> If it continues working on the navio I think we can merge it.

If you have a navio+, would you mind to test this patch? I don't have it, but I build the **navio** target and now running it on BH board. Till the moment, it has been around 18 hours, things work normally, GCS connected the whole time, RCInput responds well on GCS radio cal UI. I appreciate your help!

> As for the documentation site, send pull requests to https://github.com/ArduPilot/ardupilot_wiki - maybe @rmackay9 can help you with that.

Thanks very much for the information, I will try to launch a PR. I found the page http://ardupilot.org/copter/docs/common-autopilots.html and the overview page of each board beneath it. I would start from here:)

lucasdemarchi commented:
I updated the PR by pushing to your branch to trigger a new build. Not sure why github closed the PR and doesn't le me reopen it.

OXINARF commented:
@lucasdemarchi Not sure what you did, but the branch where the PR is from is https://github.com/bithollow/ardupilot/tree/master and that's just behind our master, so there isn't anything to PR and hence why, I guess, GitHub doesn't allow re-opening.

onceme commented:
@auturgy Thanks for the info!

@lucasdemarchi @OXINARF It is really strange, I pushed a new commit on bithollow/ardupilot/tree/master like 20 hours ago, now it's gone... I'll push again soon.

onceme commented:
I pushed again, now we can see the commit here again: https://github.com/bithollow/ardupilot/tree/master

Since @lucasdemarchi launched https://github.com/ArduPilot/ardupilot/pull/8967 to continue, let's switch to there. Thanks!
